File: sdk/javascript/src/api/knowledge.ts
/**
 * Knowledge API for Builder Engine SDK
 */

import type { HttpClient } from '../utils/http';
import type {
  KnowledgeBase,
  CreateKnowledgeBaseRequest,
  KnowledgeSearchResult,
  RequestOptions,
  PaginatedResponse,
  Metadata,
} from '../types';

export class KnowledgeAPI {
  constructor(private readonly http: HttpClient) {}

  /**
   * Create a new knowledge base
   */
  async create(
    request: CreateKnowledgeBaseRequest,
    options?: RequestOptions
  ): Promise<KnowledgeBase> {
    const response = await this.http.post<KnowledgeBase>(
      '/v1/knowledge',
      {
        agent_id: request.agentId,
        name: request.name,
        description: request.description,
      },
      options
    );
    return this.transformKnowledgeBase(response);
  }

  /**
   * Get a knowledge base by ID
   */
  async get(kbId: string, options?: RequestOptions): Promise<KnowledgeBase> {
    const response = await this.http.get<KnowledgeBase>(
      `/v1/knowledge/${kbId}`,
      undefined,
      options
    );
    return this.transformKnowledgeBase(response);
  }

  /**
   * List knowledge bases
   */
  async list(
    params?: {
      agentId?: string;
      limit?: number;
      offset?: number;
    },
    options?: RequestOptions
  ): Promise<PaginatedResponse<KnowledgeBase>> {
    const response = await this.http.get<{
      knowledge_bases: KnowledgeBase[];
      total: number;
      limit: number;
      offset: number;
    }>(
      '/v1/knowledge',
      {
        agent_id: params?.agentId,
        limit: params?.limit,
        offset: params?.offset,
      },
      options
    );

    return {
      items: response.knowledge_bases.map((kb) => this.transformKnowledgeBase(kb)),
      total: response.total,
      limit: response.limit,
      offset: response.offset,
      hasMore: response.offset + response.knowledge_bases.length < response.total,
    };
  }

  /**
   * Delete a knowledge base
   */
  async delete(kbId: string, options?: RequestOptions): Promise<void> {
    await this.http.delete(`/v1/knowledge/${kbId}`, options);
  }

  /**
   * Add text to knowledge base
   */
  async addText(
    kbId: string,
    text: string,
    metadata?: Metadata,
    options?: RequestOptions
  ): Promise<{ documentId: string; chunks: number }> {
    const response = await this.http.post<{ document_id: string; chunks: number }>(
      `/v1/knowledge/${kbId}/documents`,
      { text, metadata: metadata || {} },
      options
    );
    return {
      documentId: response.document_id,
      chunks: response.chunks,
    };
  }

  /**
   * Add URL content to knowledge base
   */
  async addUrl(
    kbId: string,
    url: string,
    options?: {
      crawlLinks?: boolean;
      maxDepth?: number;
      metadata?: Metadata;
    } & RequestOptions
  ): Promise<{ documentId: string; status: string }> {
    const response = await this.http.post<{ document_id: string; status: string }>(
      `/v1/knowledge/${kbId}/urls`,
      {
        url,
        crawl_links: options?.crawlLinks ?? false,
        max_depth: options?.maxDepth ?? 1,
        metadata: options?.metadata || {},
      },
      options
    );
    return {
      documentId: response.document_id,
      status: response.status,
    };
  }

  /**
   * Add file to knowledge base (not implemented - requires multipart upload)
   */
  async addFile(
    _kbId: string,
    _file: File | Blob,
    _options?: RequestOptions
  ): Promise<never> {
    throw new Error('File upload not implemented. Use addText or addUrl instead.');
  }

  /**
   * Search knowledge base
   */
  async search(
    kbId: string,
    query: string,
    options?: {
      limit?: number;
      threshold?: number;
      filter?: Metadata;
    } & RequestOptions
  ): Promise<KnowledgeSearchResult[]> {
    const response = await this.http.get<{ results: KnowledgeSearchResult[] }>(
      `/v1/knowledge/${kbId}/search`,
      {
        query,
        limit: options?.limit ?? 5,
        threshold: options?.threshold,
        filter: options?.filter ? JSON.stringify(options.filter) : undefined,
      },
      options
    );
    return response.results;
  }

  /**
   * Trigger sync for knowledge base
   */
  async sync(kbId: string, options?: RequestOptions): Promise<{ status: string }> {
    return this.http.post<{ status: string }>(
      `/v1/knowledge/${kbId}/sync`,
      undefined,
      options
    );
  }

  /**
   * Get documents in a knowledge base
   */
  async listDocuments(
    kbId: string,
    params?: {
      limit?: number;
      offset?: number;
    },
    options?: RequestOptions
  ): Promise<
    PaginatedResponse<{
      id: string;
      sourceType: string;
      sourceUrl?: string;
      chunkCount: number;
      createdAt: string;
      metadata: Metadata;
    }>
  > {
    const response = await this.http.get<{
      documents: Array<{
        id: string;
        source_type: string;
        source_url?: string;
        chunk_count: number;
        created_at: string;
        metadata: Metadata;
      }>;
      total: number;
      limit: number;
      offset: number;
    }>(`/v1/knowledge/${kbId}/documents`, params, options);

    return {
      items: response.documents.map((doc) => ({
        id: doc.id,
        sourceType: doc.source_type,
        sourceUrl: doc.source_url,
        chunkCount: doc.chunk_count,
        createdAt: doc.created_at,
        metadata: doc.metadata,
      })),
      total: response.total,
      limit: response.limit,
      offset: response.offset,
      hasMore: response.offset + response.documents.length < response.total,
    };
  }

  /**
   * Delete a document from knowledge base
   */
  async deleteDocument(
    kbId: string,
    documentId: string,
    options?: RequestOptions
  ): Promise<void> {
    await this.http.delete(`/v1/knowledge/${kbId}/documents/${documentId}`, options);
  }

  /**
   * Get knowledge base statistics
   */
  async getStats(
    kbId: string,
    options?: RequestOptions
  ): Promise<{
    totalDocuments: number;
    totalChunks: number;
    totalTokens: number;
    lastUpdated: string;
  }> {
    const response = await this.http.get<{
      total_documents: number;
      total_chunks: number;
      total_tokens: number;
      last_updated: string;
    }>(`/v1/knowledge/${kbId}/stats`, undefined, options);

    return {
      totalDocuments: response.total_documents,
      totalChunks: response.total_chunks,
      totalTokens: response.total_tokens,
      lastUpdated: response.last_updated,
    };
  }

  // Private helpers

  private transformKnowledgeBase(data: Record<string, unknown>): KnowledgeBase {
    return {
      id: data.id as string,
      agentId: (data.agentId || data.agent_id) as string,
      name: data.name as string,
      description: data.description as string | undefined,
      sourceType: (data.sourceType || data.source_type) as KnowledgeBase['sourceType'],
      chunkCount: (data.chunkCount || data.chunk_count) as number,
      status: data.status as KnowledgeBase['status'],
      createdAt: (data.createdAt || data.created_at) as string,
      updatedAt: (data.updatedAt || data.updated_at) as string,
    };
  }
}

File: sdk/python/bvrai_sdk/knowledge.py
"""Knowledge API for Builder Engine."""

from typing import Optional, Dict, List, Any, TYPE_CHECKING
from dataclasses import dataclass, field

if TYPE_CHECKING:
    from bvrai_sdk.client import BvraiClient


@dataclass
class KnowledgeBase:
    """A knowledge base for an agent."""
    id: str
    agent_id: str
    name: str
    description: Optional[str] = None
    source_type: str = "text"
    chunk_count: int = 0
    status: str = "pending"
    created_at: Optional[str] = None
    updated_at: Optional[str] = None

    @classmethod
    def from_dict(cls, data: Dict[str, Any]) -> "KnowledgeBase":
        return cls(
            id=data["id"],
            agent_id=data["agent_id"],
            name=data["name"],
            description=data.get("description"),
            source_type=data.get("source_type", "text"),
            chunk_count=data.get("chunk_count", 0),
            status=data.get("status", "pending"),
            created_at=data.get("created_at"),
            updated_at=data.get("updated_at"),
        )


class KnowledgeAPI:
    """
    Knowledge API client.

    Manage knowledge bases for agent RAG.
    """

    def __init__(self, client: "BvraiClient"):
        self._client = client

    async def create(
        self,
        agent_id: str,
        name: str,
        description: Optional[str] = None,
    ) -> KnowledgeBase:
        """Create a new knowledge base."""
        data = {
            "agent_id": agent_id,
            "name": name,
            "description": description,
        }
        response = await self._client.post("/v1/knowledge", data=data)
        return KnowledgeBase.from_dict(response)

    async def get(self, kb_id: str) -> KnowledgeBase:
        """Get a knowledge base."""
        response = await self._client.get(f"/v1/knowledge/{kb_id}")
        return KnowledgeBase.from_dict(response)

    async def list(self, agent_id: Optional[str] = None) -> List[KnowledgeBase]:
        """List knowledge bases."""
        params = {}
        if agent_id:
            params["agent_id"] = agent_id
        response = await self._client.get("/v1/knowledge", params=params)
        return [KnowledgeBase.from_dict(kb) for kb in response.get("knowledge_bases", [])]

    async def delete(self, kb_id: str) -> bool:
        """Delete a knowledge base."""
        await self._client.delete(f"/v1/knowledge/{kb_id}")
        return True

    async def add_text(
        self,
        kb_id: str,
        text: str,
        metadata: Optional[Dict[str, Any]] = None,
    ) -> Dict[str, Any]:
        """Add text to knowledge base."""
        data = {"text": text, "metadata": metadata or {}}
        return await self._client.post(f"/v1/knowledge/{kb_id}/documents", data=data)

    async def add_url(
        self,
        kb_id: str,
        url: str,
        crawl_links: bool = False,
    ) -> Dict[str, Any]:
        """Add URL content to knowledge base."""
        data = {"url": url, "crawl_links": crawl_links}
        return await self._client.post(f"/v1/knowledge/{kb_id}/urls", data=data)

    async def add_file(
        self,
        kb_id: str,
        file_path: str,
    ) -> Dict[str, Any]:
        """Add file to knowledge base."""
        # File upload would use multipart form
        raise NotImplementedError("File upload requires multipart support")

    async def search(
        self,
        kb_id: str,
        query: str,
        limit: int = 5,
    ) -> List[Dict[str, Any]]:
        """Search knowledge base."""
        params = {"query": query, "limit": limit}
        response = await self._client.get(f"/v1/knowledge/{kb_id}/search", params=params)
        return response.get("results", [])

    async def sync(self, kb_id: str) -> Dict[str, Any]:
        """Trigger sync for knowledge base."""
        return await self._client.post(f"/v1/knowledge/{kb_id}/sync")
